### ISSUE TITLE
isinstance type narrowing ignored when called in extra function

### ISSUE THREAD

**Issue by KerstenBreuer** [python/mypy#11475]
**Bug Report**

**To Reproduce**

I try to perform type narrowing using `isinstance` and raise an error when the wrong type is detected. This works like a charm like this:

```python
from typing import Optional

def some_function(param: Optional[str]):
    if not isinstance(param, str):
        raise TypeError()
    param_modified = param.upper
```

However, when I move the `isinstance` call into a dedicated function, type narrowing doesn't work anymore:

```python
from typing import Optional


def my_type_guard(param):
    if not isinstance(param, str):
        raise TypeError()


def some_function(param: Optional[str]):
    my_type_guard(param)
    param_modified = param.upper
```

**Expected Behavior**

In both examples, mypy should correctly infer that the param type is `str` before calling `param.upper`.

**Actual Behavior**

The mypy's output for the second example is:
``` 
error.py:11: error: Item "None" of "Optional[str]" has no attribute "upper"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: no
- Mypy configuration options from `mypy.ini` (and other config files): no
- Python version used: 3.9.7
- Operating system and version: debian buster

Thanks a lot, for your help.

**Comment by sobolevn** [comment]
This is not a bug, this is more like a "new feature"

**Comment by sobolevn** [comment]
For now you can use:

```python
from typing_extensions import TypeGuard
from typing import Optional

def my_type_guard(param: object) -> TypeGuard[str]:
    if not isinstance(param, str):
        raise TypeError()
    return True


def some_function(param: Optional[str]):
    assert my_type_guard(param)
    reveal_type(param)  # N:  Revealed type is "builtins.str"
```

**Comment by JelleZijlstra** [comment]
Yes, this is a case for TypeGuard.

**Comment by KerstenBreuer** [comment]
Thanks a lot @JelleZijlstra.
